### PR TITLE
[5.8] Upgrade Carbon to version 2 and allow CarbonImmutable

### PR DESCRIPTION
**Pull request by kylekatarnls**
Hi,

This PR is not ready to be merged yet, this is a proof of concept to be discussed. This is about upgrading to Carbon 2 and as Carbon 2 comes with a factory, it could allow Laravel users to choose `CarbonImmutable` as default class for dates (immutability seems very asked). Or it would allow to change in a easier way the class/engine to use for date with anything (jessengers/date, chronos, or simply a custom user sub-class).

It would be done with:
```php
app(\Carbon\Factory::class)->setClassName(\Carbon\CarbonImmutable::class);
```

After this code, any `now()`, `today()` or inner Carbon creations would create CarbonImmutable instances.

I targeted 5.8, tell me if you think it should be 5.7 or else.

If you think it's the right way to do it. I could go further with tests, doc and maybe a dedicated place in config to set date class and settings.

### CODE REVIEW

**Comment by browner12** [comment]
I think 5.8 is definitely the right choice. 5.7 is coming out fairly soon, and Carbon v2 is still only in beta.

**Comment by deleugpn** [comment]
Having Immutable Dates in 5.7 would be the cherry in the pie.

File: composer.json [hunk 1]
```diff
@@ -23,7 +23,7 @@
         "erusev/parsedown": "^1.7",
         "league/flysystem": "^1.0.8",
         "monolog/monolog": "^1.12",
-        "nesbot/carbon": "^1.26.3",
+        "nesbot/carbon": "^2.0.0-beta.5",
```
Comment (GrahamCampbell) [inline composer.json]: Just `^2.0` would be fine.

Comment (GrahamCampbell) [reply composer.json]: Please also update the other composer.json files in this project.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by kylekatarnls** [comment]
I thought StyleCI expected alphabetical order for imports as my IDE set them, but seems not. Someone can explain to me the expected order? https://github.styleci.io/analyses/8AEyl9

**Comment by X-Coder264** [comment]
Nope, Laravel style is to sort them by length first. And StyleCI clearly depicts that by showing what needs to be placed where.

**Comment by X-Coder264** [comment]
You can apply rules for the repo on the StyleCI website if you are the owner of the repo. So there is no need for the styleci.yml to be in the repo.

**Comment by crynobone** [comment]
Are we going to start using `app()` everywhere in the framework?

File: src/Illuminate/Auth/Passwords/DatabaseTokenRepository.php [hunk 2]
```diff
@@ -157,7 +161,7 @@ public function delete(CanResetPasswordContract $user)
      */
     public function deleteExpired()
     {
-        $expiredAt = Carbon::now()->subSeconds($this->expires);
+        $expiredAt = app(Factory::class)->now()->subSeconds($this->expires);
 
         $this->getTable()->where('created_at', '<', $expiredAt)->delete();
```
Comment (crynobone) [inline src/Illuminate/Auth/Passwords/DatabaseTokenRepository.php]: Auth component doesn't load `app()` which is under `Illuminate\Foundation`, nor does it require `illuminate/container`.

File: src/Illuminate/Cache/Repository.php [hunk 3]
```diff
@@ -554,7 +554,8 @@ protected function getMinutes($duration)
         $duration = $this->parseDateInterval($duration);
 
         if ($duration instanceof DateTimeInterface) {
-            $duration = Carbon::now()->diffInSeconds(Carbon::createFromTimestamp($duration->getTimestamp()), false) / 60;
+            $dateFactory = app(Factory::class);
+            $duration = $dateFactory->now()->diffInSeconds($dateFactory->createFromTimestamp($duration->getTimestamp()), false) / 60;
```
Comment (crynobone) [inline src/Illuminate/Cache/Repository.php]: Cache component doesn't load `app()` which is under `Illuminate\Foundation`, nor does it require `illuminate/container`.

File: src/Illuminate/Console/Scheduling/Event.php [hunk 4]
```diff
@@ -687,11 +690,11 @@ public function getSummaryForDisplay()
...
      * @param  bool  $allowCurrentDate
-     * @return \Illuminate\Support\Carbon
+     * @return \Carbon\Carbon
      */
     public function nextRunDate($currentTime = 'now', $nth = 0, $allowCurrentDate = false)
     {
-        return Carbon::instance(CronExpression::factory(
+        return app(Factory::class)->instance(CronExpression::factory(
             $this->getExpression()
         )->getNextRunDate($currentTime, $nth, $allowCurrentDate, $this->timezone));
```
Comment (crynobone) [inline src/Illuminate/Console/Scheduling/Event.php]: Console component doesn't load `app()` which is under `Illuminate\Foundation`, nor does it require `illuminate/container`.

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 5]
```diff
@@ -727,22 +728,24 @@ protected function asDate($value)
...
     {
         // If this value is already a Carbon instance, we shall just return it as is.
         // This prevents us having to re-instantiate a Carbon instance when we know
         // it already is one, which wouldn't be fulfilled by the DateTime check.
-        if ($value instanceof Carbon) {
+        if ($value instanceof CarbonInterface) {
             return $value;
         }
 
+        $dateFactory = app(Factory::class);
```
Comment (crynobone) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Database component doesn't load `app()` which is under `Illuminate\Foundation`.

**Comment by X-Coder264** [comment]
@crynobone Good question. I think we should use `app()` as little as possible or even better avoid it completely.

File: src/Illuminate/Filesystem/FilesystemAdapter.php [hunk 6]
```diff
@@ -548,7 +548,7 @@ public function getAwsTemporaryUrl($adapter, $path, $expiration, $options)
     public function getRackspaceTemporaryUrl($adapter, $path, $expiration, $options)
     {
         return $adapter->getContainer()->getObject($path)->getTemporaryUrl(
-            Carbon::now()->diffInSeconds($expiration),
+            app(Factory::class)->now()->diffInSeconds($expiration),
```
Comment (crynobone) [inline src/Illuminate/Filesystem/FilesystemAdapter.php]: Filesystem component doesn't load `app()` which is under `Illuminate\Foundation`, nor does it require `illuminate/container`.

**Comment by crynobone** [comment]
This would fine on the framework level, but once you need to pull in a single component none will work with this changes.

**Comment by kylekatarnls** [comment]
OK, but what do you propose so? Remember the desired feature: the user should be able to set somewhere the class he want for any date accessible in at application level.

So it may not be needed everywhere. Carbon instances used then destroyed and never returned could still be created the old way. But if the user would have chosen CarbonImmutable as date type, he certainly would expect to get CarbonImmutable for datetime columns in Eloquent models.

**Comment by crynobone** [comment]
> Remember the desired feature: the user should be able to set somewhere the class he want for any date accessible at application level.

Who desire this feature? What's the advantage of requiring `nesbot/carbon` when Laravel plan to support other date libraries?

I don't see any advantage on resolving factory when a class only generate a carbon instance and use the output. The only component that might heavily take advantage of Immutable is the database component, and maybe for queues scheduling.

File: src/Illuminate/Auth/Notifications/VerifyEmail.php [hunk 7]
```diff
@@ -59,7 +59,7 @@ public function toMail($notifiable)
     protected function verificationUrl($notifiable)
     {
         return URL::temporarySignedRoute(
-            'verification.verify', Carbon::now()->addMinutes(60), ['id' => $notifiable->getKey()]
+            'verification.verify', app(Factory::class)->now()->addMinutes(60), ['id' => $notifiable->getKey()]
```
Comment (crynobone) [inline src/Illuminate/Auth/Notifications/VerifyEmail.php]: This is just going to create a "60 minutes datetime from now", a plain `Carbon::now()->addMinutes(60)` should work fine.

File: src/Illuminate/Auth/Passwords/DatabaseTokenRepository.php [hunk 8]
```diff
@@ -107,7 +107,11 @@ protected function deleteExisting(CanResetPasswordContract $user)
      */
     protected function getPayload($email, $token)
     {
-        return ['email' => $email, 'token' => $this->hasher->make($token), 'created_at' => new Carbon];
+        return [
+            'email' => $email,
+            'token' => $this->hasher->make($token),
+            'created_at' => app(Factory::class)->now(),
```
Comment (crynobone) [inline src/Illuminate/Auth/Passwords/DatabaseTokenRepository.php]: This is just used to store "now" as `created_at` for `reset_passwords`, `Carbon::now()` should be more than enough.

File: src/Illuminate/Auth/Passwords/DatabaseTokenRepository.php [hunk 9]
```diff
@@ -136,7 +140,7 @@ public function exists(CanResetPasswordContract $user, $token)
      */
     protected function tokenExpired($createdAt)
     {
-        return Carbon::parse($createdAt)->addSeconds($this->expires)->isPast();
+        return app(Factory::class)->parse($createdAt)->addSeconds($this->expires)->isPast();
```
Comment (crynobone) [inline src/Illuminate/Auth/Passwords/DatabaseTokenRepository.php]: This just to check if the token has expired. Why do we need to make it complicated?

**Comment by kylekatarnls** [comment]
Thanks all for your help. I will have a second iteration on this according to your recommendations:
- target public dates only
- use stuff from illuminate/support, so we get no inner dependencies change

**Comment by kylekatarnls** [comment]
Yes, it's exactly what I planned and why I close this one. Carbon project will not handle this because we want to abandon the facade pattern (mostly static setters), then each third-party library could get its own factory with its own settings with no impact for the app. So I would create this facade in the Illuminate\Support namespace to carry the factory dedicated to Laravel dates exposed to the user.

**Comment by crynobone** [comment]
We don't need `immutable` and `mutable` options everywhere in the framework. Most of the usage above were done to utilise `Carbon::setTestNow()` so we can utilise Carbon for testing. Otherwise `time()` or `new DateTime()` would be sufficient.

Yes there is a request or needs for `immutable`, but as I said earlier maybe on the database components specifically.

**Pull request closed without merging**